src/data-model-providers/codegen/ServerClusterInterfaceRegistry.h

```diff
@@ -0,0 +1,132 @@
+/*
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/ConcreteClusterPath.h>
+#include <app/server-cluster/ServerClusterInterface.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+
+namespace chip {
+namespace app {
+
+/// Represents an entry in the server cluster interface registry for
+/// a specific interface.
+///
+/// In practice this is a single-linked list element.
+struct ServerClusterRegistration
+{
+    // A single-linked list of clusters registered for the given `endpointId`
+    ServerClusterInterface * serverClusterInterface = nullptr;
+    ServerClusterRegistration * next                = nullptr;
+
+    constexpr ServerClusterRegistration() = default;
```

Comment:
Followup, since it will need changes to the test code too: I think we should have only one constructor here, and it should take one argument: the ServerClusterInterface.

And arguably it should take that by reference, not by value, since null is not in fact allowed...

The test code bits that want to test "already in list" can register a thing and then try registering it again.  The bits that want to test the null part... we should make null not be a thing.  ;)

Reply:
I would follow up once we use this a while. I was thinking our ember cluster code sometimes declares arrays of things so it may not be immediately obvious how to initialize these pointers outside of some `init` method (i.e. may not be able to cal lthe right constructors right away).

We could have some class that contains both instance and registration, so it can correctly construct things, however the reality is that I have not prototyped that yet so I did not want to add restrictions that would cause problems in the future.

Reply:
Will try to see the sideffects: without a default constructor, std::vector::reserve is not available. However that is maybe ok.

Reply:
Updated - test cannot reserve vectors anymore, however now we should be more null safe.